src/awkward/_v2/operations/structure/ak_concatenate.py

```diff
@@ -4,524 +4,201 @@
+    #         mergebool (bool): If True, boolean and nummeric data can be combined
+    #             into the same buffer, losing information about False vs `0` and
+    #             True vs `1`; otherwise, they are kept in separate buffers with
+    #             distinct types (using an #ak.layout.UnionArray8_64).
+    #         highlevel (bool): If True, return an #ak.Array; otherwise, return
+    #             a low-level #ak.layout.Content subclass.
+    #         behavior (None or dict): Custom #ak.behavior for the output array, if
+    #             high-level.
+
+    #     Returns an array with `arrays` concatenated. For `axis=0`, this means that
+    #     one whole array follows another. For `axis=1`, it means that the `arrays`
+    #     must have the same lengths and nested lists are each concatenated,
+    #     element for element, and similarly for deeper levels.
+    #     """
+
+    contents = [
+        ak._v2.operations.convert.to_layout(
+            x, allow_record=False if axis == 0 else True, allow_other=True
+        )
+        for x in arrays
+    ]
+    if not any(isinstance(x, (ak._v2.contents.Content,)) for x in contents):
+        raise ValueError("need at least one array to concatenate")
+
+    first_content = [x for x in contents if isinstance(x, (ak._v2.contents.Content,))][
+        0
+    ]
+
+    posaxis = first_content.axis_wrap_if_negative(axis)
+    maxdepth = max(
+        [
+            x.minmax_depth[1]
+            for x in contents
+            if isinstance(
+                x,
+                (ak._v2.contents.Content,),
+            )
+        ]
+    )
+    if not 0 <= posaxis < maxdepth:
+        raise ValueError(
+            "axis={0} is beyond the depth of this array or the depth of this array "
+            "is ambiguous".format(axis)
+        )
+    for x in contents:
+        if isinstance(x, ak._v2.contents.Content):
+            if x.axis_wrap_if_negative(axis) != posaxis:
+                raise ValueError(
+                    "arrays to concatenate do not have the same depth for negative "
+                    "axis={0}".format(axis)
+                )
+
+    if posaxis == 0:
+        contents = [
+            x
+            if isinstance(x, ak._v2.contents.Content)
+            else ak._v2.operations.convert.to_layout([x])
+            for x in contents
+        ]
+        batch = [contents[0]]
+        for x in contents[1:]:
+            if batch[-1].mergeable(x, mergebool=mergebool):
+                batch.append(x)
+            else:
+                collapsed = batch[0].mergemany(batch[1:])
+                batch = [collapsed.merge_as_union(x)]
+
+        out = batch[0].mergemany(batch[1:])
+        if isinstance(out, ak._v2.contents.unionarray.UnionArray):
+            # if isinstance(out, ak._v2._util.uniontypes):
+            out = out.simplify_uniontype(merge=merge, mergebool=mergebool)
+
+    else:
+
+        def getfunction(inputs, depth):
```

Comment:
This can be replaced with an `action` that passes the depth down through a `depth_context`, following the example in the v2 `from_regular`. That one uses `recursively_apply`, rather than `broadcast_and_apply`, but the two interfaces have been made the same now. There are also examples of v2 `broadcast_and_apply` in `_connect/numpy.py`.